v2/snowflake-common/src/main/java/com/google/cloud/teleport/v2/snowflake/util/DecryptUtil.java

```diff
@@ -0,0 +1,159 @@
+
+	private static final Pattern ENCRYPT_PATTERN = Pattern.compile("\\$\\{(.*?)}");
+
+	/** The log to output status messages to. */
+	private static final Logger LOG = LoggerFactory.getLogger(DecryptUtil.class);
+
+	/**
+	 * Uses the GCP KMS client to decrypt an encrypted value using a KMS key of the
+	 * form
+	 * projects/{gcp_project}/locations/{key_region}/keyRings/{key_ring}/cryptoKeys/{kms_key_name}.
+	 * The encrypted value should be a base64 encrypted string which has been
+	 * encrypted using the KMS encrypt API call. See <a href=
+	 * "https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings.cryptoKeys/encrypt">
+	 * this KMS API Encrypt Link</a>.
+	 * 
+	 * @param encryptedValue base64 encrypted string
+	 * @param kmsKey encryption key
+	 * @return decrypted value
+	 * @throws IOException
+	 */
+
+	private String kmsKey;
+	private KeyManagementServiceClient keyManagementServiceClient;
+
+	public DecryptUtil(String kmsKey) throws IOException {
+		if (Strings.isNullOrEmpty(kmsKey)|| !validateKmsKey(kmsKey)){
+			IllegalArgumentException exception = new IllegalArgumentException("Provided KMS Key %s is invalid");
+			throw new RuntimeException(exception);
+		}
+		this.kmsKey = kmsKey;
+		this.keyManagementServiceClient = KeyManagementServiceClient.create();
+	}
+
+	private String decryptWithKMS(String encryptedValue) throws IOException {
+
+		byte[] cipherText = Base64.getDecoder().decode(encryptedValue.getBytes("UTF-8"));
+
+		// Decrypt the ciphertext with Cloud KMS.
+		DecryptResponse decryptResponse = this.keyManagementServiceClient.decrypt(this.kmsKey, ByteString.copyFrom(cipherText));
+
+		// Extract the plaintext from the response.
+		return new String(decryptResponse.getPlaintext().toByteArray());
+	}
+
+	/**
+	 * Validates that kms key should be in the following format:
+	 * projects/{gcp_project}/locations/{key_region}/keyRings/{key_ring}/cryptoKeys/
+	 * {kms_key_name}.
+	 * 
+	 * @param kmsKey the encryption key
+	 * @return true if key matches the pattern above else returns false
+	 */
+	private static boolean validateKmsKey(String kmsKey) {
+		return KEY_NAME_PATTERN.matcher(kmsKey).matches();
+	}
+
+	/**
+	 * Checks if pipeline parameter is encrypted. If pipeline parameter is
+	 * encrypted, returns the decrypted value using kms encryption key otherwise
+	 * returns the raw value as is.
+	 * 
+	 * @param pipelineOption runtime parameter passed to pipeline
+	 * @return decrypted value if pipeline parameter is encrypted otherwise returns
+	 *         the raw value as is.
+	 */
+	public String decryptIfRequired(ValueProvider pipelineOption) {
+		if (pipelineOption.isAccessible()) {
+			String pipelineOptionValue = (String) pipelineOption.get();
+			if (!Strings.isNullOrEmpty(pipelineOptionValue) && isEncrypted(pipelineOptionValue)) {
```

Comment:
Below code improves readability and avoid pattern matching twice in case option is encrypted:
```
public String decryptIfRequired(ValueProvider<String> pipelineOption) throws IOException {
        if (pipelineOption.isAccessible() && !Strings.isNullOrEmpty(pipelineOption.get())) {
            String pipelineOptionValue = pipelineOption.get();
            Matcher match = ENCRYPT_PATTERN.matcher(pipelineOptionValue);
            return match.find() ? decryptWithKMS(match.group(1)) : pipelineOptionValue;
        }
        return null;
    }
```
Above code also eliminates removing additional methods